### PR TITLE
Windows installer for storage nodes

### PR DESCRIPTION
**Pull request by kaloyan-raev**
What: Scripts and descriptors for building MSI package using the Wix Toolset.

To build the MSI package:
1. You need a Windows system/VM.
1. Install latest WiX Toolset 3.x from https://wixtoolset.org/.
1. Download the storage node binary for Windows, extract it, rename it to `storagenode.exe`, and copy it to the `installer/windows` folder.
1. Install Visual Studio 2019 Community: https://visualstudio.microsoft.com/downloads/.
1. Install [Wix Toolset Visual Studio 2019 Extension](https://marketplace.visualstudio.com/items?itemName=WixToolset.WixToolsetVisualStudio2019Extension).
1. Open the `installer/windows/windows.sln` file in Visual Studio.
1. Build the solution: Build > Build Solution.
1. Find the MSI at `installer\windows\bin\Debug\storagenode.msi`.

This is work-in-progress. So far, the installer:
- [x] Displays AGPL license
- [x] Allows custom install folder (default is C:\Program Files\Storj\Storage Node)
- [x] Installs the storagenode.exe
- [x] Dialog for selecting identity dir
- [x] Dialog for entering wallet and email addresses
- [x] Dialog for entering public address
- [x] Dialog for entering storage dir, available storage and bandwidth
- [x] Custom actions for validating the config properties
- [x] Custom action for generating config.yaml
- [x] Register storagenode.exe as a Window Service
- [x] Firewall exception for port 28967
- [x] Shortcut for opening the dashboard
- [x] Branding images

Why: We want to support Windows Home. Jira issue: https://storjlabs.atlassian.net/browse/V3-2430

Please describe the tests:
 - Test 1: No automatic tests
 
Please describe the performance impact: Not relevant

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: installer/windows/agpl-3.0.rtf [hunk 1]
```diff
@@ -0,0 +1,277 @@
+{\rtf1\ansi\deff3\adeflang1025
```
Comment (egonelbre) [inline installer/windows/agpl-3.0.rtf]: I'm not sure whether we need to include AGPL here. We're not importing any of their code here.

File: installer/windows/CustomInstallDir.wxs [hunk 2]
```diff
@@ -0,0 +1,123 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Customized to collect user input for storage node config.
```
Comment (mniewrzal) [inline installer/windows/CustomInstallDir.wxs]: Are these comments up to date?

File: installer/windows/CustomInstallDir.wxs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            <Publish Dialog="AllocatedBandwidthConfigDlg" Control="Next" Event="SpawnDialog" Value="ValidationErrorDlg" Order="3"><![CDATA[STORJ_BANDWIDTH_VALID<>"1"]]></Publish>
+            <Publish Dialog="AllocatedBandwidthConfigDlg" Control="Next" Event="NewDialog" Value="VerifyReadyDlg" Order="4">STORJ_BANDWIDTH_VALID="1"</Publish>
+          
+            <Publish Dialog="VerifyReadyDlg" Control="Back" Event="NewDialog" Value="AllocatedBandwidthConfigDlg" Order="1">NOT Installed</Publish>
+            <Publish Dialog="VerifyReadyDlg" Control="Back" Event="NewDialog" Value="MaintenanceTypeDlg" Order="2">Installed AND NOT PATCH</Publish>
+            <Publish Dialog="VerifyReadyDlg" Control="Back" Event="NewDialog" Value="WelcomeDlg" Order="2">Installed AND PATCH</Publish>
+
+            <Publish Dialog="MaintenanceWelcomeDlg" Control="Next" Event="NewDialog" Value="MaintenanceTypeDlg">1</Publish>
+
+            <Publish Dialog="MaintenanceTypeDlg" Control="RepairButton" Event="NewDialog" Value="VerifyReadyDlg">1</Publish>
```
Comment (mniewrzal) [inline installer/windows/CustomInstallDir.wxs]: Maybe we need to disable this button for now?

File: installer/windows/EmailConfig.wxs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+        <Control Id="Description" Type="Text" X="25" Y="23" Width="280" Height="15" Transparent="yes" NoPrefix="yes" Text="Enter your e-mail address." />
+        <Control Id="Title" Type="Text" X="15" Y="6" Width="200" Height="15" Transparent="yes" NoPrefix="yes" Text="{\WixUI_Font_Title}Operator Information" />
+        <Control Id="BannerBitmap" Type="Bitmap" X="0" Y="0" Width="370" Height="44" TabSkip="no" Text="!(loc.InstallDirDlgBannerBitmap)" />
+        <Control Id="BannerLine" Type="Line" X="0" Y="44" Width="370" Height="0" />
+        <Control Id="BottomLine" Type="Line" X="0" Y="234" Width="370" Height="0" />
+
+        <Control Id="EmailLabel" Type="Text" X="20" Y="60" Width="320" Height="16" NoPrefix="yes" Text="E-mail address:" />
+        <Control Id="Email" Type="Edit" Property="STORJ_EMAIL" X="20" Y="100" Width="320" Height="18" />
+        <Control Id="WalletDesc" Type="Text" X="20" Y="150" Width="320" Height="48" NoPrefix="yes" Text="Optional. We will notify you on this email address when a new version has been released." />
```
Comment (mniewrzal) [inline installer/windows/EmailConfig.wxs]: invalid id `WalletDesc`

Comment (mniewrzal) [reply installer/windows/EmailConfig.wxs]: nit: I'm also thinking maybe moving `Optional` to `EmailLabel` would be more readable?

File: installer/windows/PublicAddressConfig.wxs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        <Control Id="BannerBitmap" Type="Bitmap" X="0" Y="0" Width="370" Height="44" TabSkip="no" Text="!(loc.InstallDirDlgBannerBitmap)" />
+        <Control Id="BannerLine" Type="Line" X="0" Y="44" Width="370" Height="0" />
+        <Control Id="BottomLine" Type="Line" X="0" Y="234" Width="370" Height="0" />
+
+        <Control Id="PublicAddressLabel" Type="Text" X="20" Y="60" Width="320" Height="16" NoPrefix="yes" Text="External address (&lt;ip&gt;:&lt;port&gt;):" />
+        <Control Id="PublicAddress" Type="Edit" Property="STORJ_PUBLIC_ADDRESS" X="20" Y="100" Width="320" Height="18" />
+        <Control Id="PublicAddressDesc" Type="Text" X="20" Y="150" Width="320" Height="32" NoPrefix="yes" Text="The storage node will listen for requests on port 28967. You must configure your router to make this port available for public access. When ready, enter the external IP address or the DDNS you configured and the port your opened on your router." />
+        <Control Id="IdentityHowto" Type="Hyperlink" X="20" Y="190" Width="320" Height="16">
```
Comment (mniewrzal) [inline installer/windows/PublicAddressConfig.wxs]: Invalid `Id` `IdentityHowto`

File: installer/windows/build.bat [hunk 6]
```diff
@@ -0,0 +1,3 @@
+candle.exe -arch x64 Product.wxs CustomInstallDir.wxs IdentityDir.wxs OperatorConfig.wxs
+
+light.exe -o storagenode.msi -ext WixUIExtension Product.wixobj CustomInstallDir.wixobj IdentityDir.wixobj  OperatorConfig.wixobj
```
Comment (mniewrzal) [inline installer/windows/build.bat]: This script is not working right now for me:
```
C:\Users\IEUser\git\storj\installer\windows>light.exe -o storagenode.msi -ext WixUIExtension Product.wixobj CustomInstallDir.wixobj IdentityDir.wixobj  OperatorConfig.wixobj
light.exe : error LGHT0103 : The system cannot find the file 'Product.wixobj' with type 'Source'.
```

Comment (kaloyan-raev) [reply installer/windows/build.bat]: I will remove it. It worked at the beginning, when the project was simpler, before start using Visual Studio.

File: installer/windows/CustomInstallDir.wxs (hunk 2 above)
Comment (kaloyan-raev) [reply installer/windows/CustomInstallDir.wxs]: I updated the first-time sequence.

File: installer/windows/EmailConfig.wxs (hunk 4 above)
Comment (kaloyan-raev) [reply installer/windows/EmailConfig.wxs]: Both fixed.

File: installer/windows/PublicAddressConfig.wxs (hunk 5 above)
Comment (kaloyan-raev) [reply installer/windows/PublicAddressConfig.wxs]: Fixed.

File: installer/windows/CustomInstallDir.wxs (hunk 3 above)
Comment (kaloyan-raev) [reply installer/windows/CustomInstallDir.wxs]: Good idea. Done.

**Comment by bryanchriswhite** [comment]
Lets make sure we copy the instructions in the description to the wiki or something.

**Pull request merged**